src/io/GeoJSONReader.cpp

```diff
@@ -45,210 +45,268 @@ GeoJSONReader::GeoJSONReader(): GeoJSONReader(*(GeometryFactory::getDefaultInsta
         throw  ParseException("Expected two coordinates found one");
-    } else if (coords.size() > 2) {
+    }
+    else if (coords.size() > 2) {
         throw  ParseException("Expected two coordinates found more than two");
-    } else {
+    }
+    else {
         return geom::Coordinate {coords[0], coords[1]};
     }
 }
 
-std::unique_ptr<geom::Point> GeoJSONReader::readPoint(const geos_nlohmann::json& j) {
-    auto coords = j["coordinates"].get<std::vector<double>>();
+std::unique_ptr<geom::Point> GeoJSONReader::readPoint(
+    const geos_nlohmann::json& j) const
+{
+    const auto& coords = j["coordinates"].get<std::vector<double>>();
     if (coords.size() == 1) {
         throw  ParseException("Expected two coordinates found one");
-    } else if (coords.size() < 2) {
+    }
+    else if (coords.size() < 2) {
         return geometryFactory.createPoint(2);
-    } else {
-        geom::Coordinate coord = readCoordinate(coords);
+    }
+    else {
+        const geom::Coordinate& coord = readCoordinate(coords);
         return std::unique_ptr<geom::Point>(geometryFactory.createPoint(coord));
     }
 }
 
-std::unique_ptr<geom::LineString> GeoJSONReader::readLineString(const geos_nlohmann::json& j) {
-    auto coords = j["coordinates"].get<std::vector<std::vector<double>>>();
+std::unique_ptr<geom::LineString> GeoJSONReader::readLineString(
+    const geos_nlohmann::json& j) const
+{
+    const auto& coords = j["coordinates"].get<std::vector<std::vector<double>>>();
     std::vector<geom::Coordinate> coordinates;
-    for(const auto& coord : coords) {
-        geom::Coordinate c = readCoordinate(coord);
-        coordinates.push_back(geom::Coordinate{c.x,c.y});
+    coordinates.reserve(coords.size());
+    for (const auto& coord : coords) {
+        const geom::Coordinate& c = readCoordinate(coord);
+        coordinates.push_back(c);
     }
     auto coordinateSequence = geometryFactory.getCoordinateSequenceFactory()->create(std::move(coordinates));
     return geometryFactory.createLineString(std::move(coordinateSequence));
 }
 
-std::unique_ptr<geom::Polygon> GeoJSONReader::readPolygon(const geos_nlohmann::json& json) {
-    auto polygonCoords = json["coordinates"].get<std::vector<std::vector<std::vector<double>>>>();
+std::unique_ptr<geom::Polygon> GeoJSONReader::readPolygon(
+    const geos_nlohmann::json& json) const
+{
+    const auto& polygonCoords = json["coordinates"].get<std::vector<std::vector<std::vector<double>>>>();
     return readPolygon(polygonCoords);
 }
 
-std::unique_ptr<geom::Polygon> GeoJSONReader::readPolygon(const std::vector<std::vector<std::vector<double>>>& polygonCoords) {
+std::unique_ptr<geom::Polygon> GeoJSONReader::readPolygon(
+    const std::vector<std::vector<std::vector<double>>>& polygonCoords) const
+{
     std::unique_ptr<geom::LinearRing> shell;
     std::vector<std::unique_ptr<geom::LinearRing>> rings;
-    for(const auto& ring : polygonCoords) {
+    rings.reserve(polygonCoords.size());
```

Comment:
Do we have tests for the various cases where the ordinate arrays can be zero-length? Is reserving 0 acceptable behaviour?

Reply:
> Do we have tests for the various cases where the ordinate arrays can be zero-length? Is reserving 0 acceptable behaviour?

Only the new capacity is greater than the existing capacity will trigger memory allocation.